Debugger tool for crash analysis

thapav opened the pull request:
This patch adds crash analysis to debug assertions majorly for cortex M chipsets since the complete stack information for them is not available through the current ramdumpParser tool.

tizen-build commented:
Target : [5f08f76ce629e2be9a44475e346f40c335325308] - Code Rule Check OK.

jeongchanKim commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,178 @@
+					symbol = s[2].rstrip('\n')
+					if symbol == '_stext':
+						flash_text_start_addr = hex(int(s[0], 16))
+						#flash_text_start_addr = s[0]
```
If these lines are not needed, please remove them.

sunghan-chang commented on build/configs/rtl8721csm/scripts/rlx8721d_img2_ns.ld:
```diff
@@ -191,12 +192,14 @@ SECTIONS
 	_eronly = ABSOLUTE(.);
 
+		_stext = ABSOLUTE(.);
+		_etext = ABSOLUTE(.);
```
It looks unnecessary and duplicated. Could you check this?

jeongchanKim commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,178 @@
+		print('~~~~~~~~~~~~~~~~~~~~~~~~~~  KERNEL TEXT RANGE ~~~~~~~~~~~~~~~~~~~~~~~')
+		print('flash_text_start_addr\t=\t', flash_text_start_addr)
+		print('flash_text_end_addr\t=\t', flash_text_end_addr)
+		print('ram_text_start_addr\t=\t', ram_text_start_addr)
```
Only some boards have ram text region. Is there any way to check and print only it has?

sunghan-chang commented on os/arch/arm/src/common/up_internal.h:
```diff
@@ -255,13 +255,15 @@ EXTERN uint32_t g_dabtstackbase;	/* Initial top of data abort stack */
 EXTERN uint32_t _sbss;			/* Start of .bss */
 EXTERN uint32_t _ebss;			/* End+1 of .bss */
+EXTERN uint32_t _ram_kernel_text_start;	/* Start of .text in RAM region */
+EXTERN uint32_t _ram_kernel_text_end;	/* End of .text in RAM region  */
```
Almost boards do not use this. Is it ok for them?

jeongchanKim commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,178 @@
+
+    print('~~~~~~~~~~~~~~~~~~~~~~~~  DEBUG DISPLAY SYMBOLS ~~~~~~~~~~~~~~~~~~~~~')
+    parse_file(log_file)
+    print('~~~~~~~~~~~~~~~~~~~~~~~~  DEBUG DISPLAY SYMBOLS ~~~~~~~~~~~~~~~~~~~~~')
```
I think this line is not needed. Or we don't need to print the message in this line. Just `~~~` is enough.

jeongchanKim reviewed (commented):
I think, similar functionality is in ramdumpParser. Is there any way to merge them?
Because these code are based on assert log, so if assert log is changed, we should change both of the codes. If we have common code, we can just check only one code.
Please consider it can be or not.

tizen-build commented:
Target : [982b4d2c947b7bb8990c1127a4b88d166c15bc9e] - Code Rule Check OK.

jeongchanKim commented on tools/debug/README.md:
```diff
@@ -0,0 +1,39 @@
+Address:  0xe054f07	Symbol:  __FUNCTION__.6322
+Address:  0x10010b58	Symbol:  os_start	/home/test/tizenrt/os/include/tinyara/init.h:89
+Address:  0xe000cd4	Symbol:  app_start	/home/test/tizenrt/os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c:1318
+~~~~~~~~~~~~~~~~~~~~~~~~  DEBUG DISPLAY SYMBOLS ~~~~~~~~~~~~~~~~~~~~~
```
Please update the sample log also.

jeongchanKim commented on tools/debug/README.md:
```diff
@@ -0,0 +1,39 @@
+1. Copy crash logs.
+    First copy the crash logs to a file named log_file in <root/tizenrt/tools/debug/log_file>
+2. Run Script
+    $ python cmdautotest.py
```
What is cmdautotest.py?

thapav commented:
> I think, similar functionality is in ramdumpParser. Is there any way to merge them?
> Because these code are based on assert log, so if assert log is changed, we should change both of the codes. If we have common code, we can just check only one code.
> Please consider it can be or not.

Since there are few differences amongst what we are retrieving from the assert logs merging is not a good idea according to me. (We are just obtaining stack addresses to check against kernel text range whereas in ramdumpParser they need all FP, PC etc and then unwinding the call stack)
So as of now I have made a change based on cortex M. Now we can run debug tool through ramdumpParser itself. Whenever cortex M is ecountered it will redirect to the debugger tool.

tizen-build commented:
Target : [d7cf42667eae0ad5a43a3c4e0e54537611e3c6f6] - Code Rule Check OK.

tizen-build commented:
Target : [56b2a0bb07cac72bc7c906e4accfdc772fe2511c] - Code Rule Check OK.

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+				if word[1] == ' R8':
+					t = word[2].split( )
+					# Last subword of word[2] contains the PC value
+					for sub_word in t:
```
Why do we need for loop here? 
We must just count the number of elements in the array t, make sure that there are 8 elements, and pick the last element. If there are less than 8, then we cannot be sure if the last element is PC or LR. So, in that case we must print that the logs are corrupted, and we are not able to determine the value of PC.

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+				if k == 1:
+					# word[1] contains the stack start address
+					stack_start_addr = int(word[1], 16)
+					k = 0
```
What is the use of this logic? Why do we need the stack start address?

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+					stack_start_addr = int(word[1], 16)
+					k = 0
+				for sub_word in t:
+					stack_table.append((stack_start_addr,int(sub_word,16)))
```
Stack start address is stored in stack_table, but never used.

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+					stack_start_addr = stack_start_addr + 4
+
+					# Check if the stack address lies in kernel text address range
+					if (is_kernel_text_address(hex(stack_table[i][1]))):
```
Why do we need to have a stack_table if we are calling print_symbol() for every symbol using this for loop?

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+					# word[1] contains the stack start address
+					stack_start_addr = int(word[1], 16)
+					k = 0
+				for sub_word in t:
```
We need to check if the length of the subword is exactly 8 characters. Else it is an invalid address and we need to skip it.

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+		mid = (l + h) / 2
+
+	if (l > h):
+		print('Symbol not found in system map file')
```
IF symbol is not found, we have to print the address for which its not found.. Also, its better to put this logic if only verbose debug info is required.

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,177 @@
+	# Check the address
+	if (address >= ram_text_start_addr and address < ram_text_end_addr) or (address >= flash_text_start_addr and address < flash_text_end_addr):
+		if debug:
+			print("{0} is kernel text address!".format(address))
```
Instead of printing this line here, we can leave it. 
While trying to print the symbol, we will either print the address and corresponding symbol. Or we print that the symbol for this address is not found.

kishore-sn commented on tools/dump_tool/ramdumpParser.py:
```diff
@@ -563,6 +563,17 @@ def main():
+	fd.close()
+
+	# Get arch family
+	if ('CONFIG_ARCH_FAMILY="armv8-m"' in data) or ('CONFIG_ARCH_FAMILY="armv7-m"' in data):
```
The ramdump parser tool is basically used to get the dump call stack as well as other information like heap, etc.. 
Our logic here is related to printing the call stack symbols in case of cortex m, where the FP is not available. So, I think we must move this logic to line 621. 

`if cortex m`
`then do our logic`
`else`
`do existing call stack logic using FP`
`...`
`continue with remaining ramdump parser logic for heap, etc.. `

kishore-sn commented on build/configs/rtl8721csm/scripts/rlx8721d_img2_ns.ld:
```diff
@@ -191,7 +192,7 @@ SECTIONS
 		*dq_rem.o (.text*)
 		*dq_addlast.o (.text*)
-		_etext = ABSOLUTE(.);
+		_ram_kernel_text_end = ABSOLUTE(.);
```
This change will result in some issues in our existing code. 
Where ever we are checking for stext and etext, we assume that the whole of kernel text is in flash. 
But, now if we are placing one part in flash and other in ram, then we need to change our logic in each place in our source code. Please make such required changes in src code and add it as a separate commit.

kishore-sn commented on tools/debug/README.md:
```diff
@@ -0,0 +1,36 @@
+*************************************************************
+
+~~~~~~~~~~~~~~~~~~~~~~~~  DEBUG DISPLAY SYMBOLS ~~~~~~~~~~~~~~~~~~~~~
+Address:  0xe005df8	Symbol:  test_func	/home/test/tizenrt/apps/examples/hello/hello_main.c:64
```
Instead of printing "Address" and "Symbol" in each line, we must give a column heading to the table. 
In fact, we need to have 4 columns as follows:
`Dump address          Symbol address         Symbol name             File name`

tizen-build commented:
Target : [0473838934f8bac59de4e751bb67d4f738886758] - Code Rule Check OK.

tizen-build commented:
Target : [bbb16fb3e9c5d85db25aa7b15ee25c738ea9eefc] - Code Rule Check OK.

tizen-build commented:
Target : [8d6eafc3aee096fa1b5c54889533b3b2e8a85d3c] - Code Rule Check OK.

jeongchanKim commented on build/configs/stm32f407-disc1/scripts/kernel-space.ld:
```diff
@@ -60,6 +60,7 @@ SECTIONS
 {
     .text : {
         _stext = ABSOLUTE(.);
+	_ram_kernel_text_start = ABSOLUTE(.);
```
Please match the alignment.

thapav replied:
already aligned.

jeongchanKim commented on os/arch/arm/src/armv7-m/up_hardfault.c:
```diff
@@ -137,14 +138,13 @@ int up_hardfault(int irq, FAR void *context, FAR void *arg)
 
-	if (((uintptr_t)pc >= (uintptr_t)&_stext && (uintptr_t)pc < (uintptr_t)&_etext) ||
-			(sched_self()->uspace &&
+	if ((is_kernel_text_space((void *)pc)) || (sched_self()->uspace &&
```
`is_kernel_text_space` checks not only FLASH but also RAM space. Please update the comment also.

jeongchanKim commented on os/arch/arm/src/armv7-m/up_hardfault.c:
```diff
@@ -137,14 +138,13 @@ int up_hardfault(int irq, FAR void *context, FAR void *arg)
 	/* SVCalls are expected only from the base, kernel FLASH region */
 
-	if ((uintptr_t)pc >= (uintptr_t)&_stext && (uintptr_t)pc < (uintptr_t)&_etext)
+	if (is_kernel_text_space((void *)pc))
```
Same as above.

jeongchanKim commented on os/arch/arm/src/common/up_checkspace.c:
```diff
@@ -38,11 +38,23 @@ extern uint32_t _sdata;
+bool is_kernel_text_space(void *addr)
+{
+	/* Check if the address lies in the kernel text section range */
+	if ((addr >= (void *)&_stext && addr <= (void *)&_etext) ||
```
Is it possible that `addr` reaches `_etext`? If not, `addr < (void *)&_etext` is enough.

jeongchanKim replied:
In that case, how about modifying `up_hardfault` also? It didn't check the equal to `us_textend`.

jeongchanKim commented on os/include/tinyara/arch.h:
```diff
@@ -2179,6 +2179,15 @@ void up_wdog_keepalive(void);
+ * Name: is_kernel_text_space
+ *
+ * Description:
+ *   Check the address is in kernel space or not
```
Please change the description like below.
```
Check the address is in kernel text space or not
```

jeongchanKim commented on tools/dump_tool/ramdumpParser.py:
```diff
@@ -618,6 +622,12 @@ def main():
 	try:
+		# Get arch family
+		if ('CONFIG_ARCH_FAMILY="armv8-m"' in data) or ('CONFIG_ARCH_FAMILY="armv7-m"' in data):
+			#If its cortex M, then run debugsymbolviewer script and return
```
Pleaes change `its` to `it's`.
And this is only working with cortex M? Cannot with Cortex R?

jeongchanKim commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,178 @@
+					t = word[2].split( )
+
+					# Check for corruption of PC value in logs
+					if(len(word[2].split( )) != 8):
```
Please add space after `if`.

jeongchanKim commented on tools/debug/README.md:
```diff
@@ -0,0 +1,37 @@
+# DebugSymbolViewer
+
+The Debug Symbol viewer will help the developer when a crash occurs.
```
For next line, please add double space.

jeongchanKim commented on tools/debug/README.md:
```diff
@@ -0,0 +1,37 @@
+It will show some call stack information when the application has faced an abort. To be precise, it will show the symbols corresponding to the stack addresses present in the kernel text address range.
+
+### Prerequisites
+Install python 3.8 & python 2.7, pyserial(at pip)
```
Both python 3.8 and 2.7 are required ? 3.8 cannot cover the 2.7?

thapav replied:
No, Python 3 is *not backward compatible*
This was one of the problems that would arise due to merging of the new tool with ramdumpParser, had already intimated the same.

abhishek-samsung commented on build/configs/stm32f407-disc1/scripts/user-space.ld:
```diff
@@ -63,6 +63,7 @@ SECTIONS
 
     .text : {
         _stext = ABSOLUTE(.);
+	_ram_kernel_text_start = ABSOLUTE(.);
```
Please check the indentation.

thapav replied:
already aligned

abhishek-samsung commented on build/configs/stm32l4r9ai-disco/scripts/kernel-space.ld:
```diff
@@ -43,6 +43,7 @@ SECTIONS
 {
     .text : {
         _stext = ABSOLUTE(.);
+	_ram_kernel_text_start = ABSOLUTE(.);
```
Please check the indentation.

thapav replied:
already aligned

abhishek-samsung commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,178 @@
+				t = word[2].split( )
+				for sub_word in t:
+					# Check for valid address, else move to the next sub_word
+					if(bytes_needed(int(sub_word,16)) != 4):
```
Please leave space after if.

dineshdas-samsung commented on build/configs/stm32l4r9ai-disco/scripts/user-space.ld:
```diff
@@ -46,6 +46,7 @@ SECTIONS
 
     .text : {
         _stext = ABSOLUTE(.);
+	_ram_kernel_text_start = ABSOLUTE(.);
```
Please check the indentation here also.

thapav replied:
already aligned

tizen-build commented:
Target : [eeb5e2b7303d160fe81a863c77282a59743bedf9] - Code Rule Check OK.

sunghan-chang commented:
@thapav There is build break on qemu/build_test.
```
LD: tinyara
/root/tizenrt/os/../build/output/libraries/libarch.a(up_hardfault.o): In function `up_hardfault':
/root/tizenrt/os/arch/arm/src/armv7-m/up_hardfault.c:147: undefined reference to `is_kernel_text_space'
Makefile:199: recipe for target '../build/output/bin/tinyara' failed
make[1]: *** [../build/output/bin/tinyara] Error 1
make[1]: Leaving directory '/root/tizenrt/os/arch/arm/src'
Makefile.unix:476: recipe for target 'pass2' failed
make: *** [pass2] Error 2
"qemu/build_test" ERROR!!!
```

tizen-build commented:
Target : [f6d39ef319902781fb2b94e4e7d4b86a378a359a] - Code Rule Check OK.

tizen-build commented:
Target : [56b28c4383cac0e85722398ac32a8e08ede0c850] - Code Rule Check OK.

kishore-sn commented on os/arch/arm/src/armv7-m/up_hardfault.c:
```diff
@@ -84,6 +85,11 @@
+#ifdef CONFIG_BUILD_FLAT
+extern uint32_t _ram_kernel_text_start;
+extern uint32_t _ram_kernel_text_end;
+#endif
```
Why do we need to use these variables directly for flat build?
Can we not use the same api?

kishore-sn commented on os/arch/arm/src/common/up_internal.h:
```diff
@@ -255,8 +255,8 @@ EXTERN uint32_t g_dabtstackbase;	/* Initial top of data abort stack */
-EXTERN uint32_t _stext;			/* Start of .text */
-EXTERN uint32_t _etext;			/* End_1 of .text + .rodata */
+EXTERN uint32_t _stext;			/* Start of .text in Flash region */
+EXTERN uint32_t _etext;			/* End_1 of .text + .rodata in Flash region */
```
We need to add the new linker variables in this file. 
This file will be included in many other places to provide extern definition for the linker variables.

kishore-sn commented on os/include/tinyara/arch.h:
```diff
@@ -2180,6 +2180,15 @@ void up_wdog_keepalive(void);
+ *   Check the address is in kernel text space or not
+ *
+ ****************************************************************************/
+bool is_kernel_text_space(void *addr);
```
Let us move this out of PROTECTED flag. We need this API at all times.

kishore-sn commented on tools/debug/README.md:
```diff
@@ -0,0 +1,37 @@
+# DebugSymbolViewer
+
+The Debug Symbol viewer will help the developer when a crash occurs.  
+It will show some call stack information when the application has faced an abort. To be precise, it will show the symbols corresponding to the stack addresses present in the kernel text address range.
```
Please mention the name and meaning of each column that we will print. 
Also, you need to highlight why we print dump address and symbol address separately. 
Dump address corresponds to an address on the stack. It might be an address of any line in the middle of a function, which was getting executed at the time of dump. 
Symbol address is the address of the first instruction of a function. Since our map file only provides symbol address, we use this to print the symbol name and line number. 
Hence, the line number printed here is the line number of the first line of the function. However, the execution might be at a later line at the time of crash. The user needs to manually check this during debugging.

kishore-sn commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,178 @@
+						#If yes, print it's corresponding symbol
+						print_symbol(stack_addr)
+
+					i = i + 1
```
If we are not using the variable i, pls remove it.

kishore-sn commented on tools/dump_tool/ramdumpParser.py:
```diff
@@ -563,6 +563,10 @@ def main():
+	# Read config information
+	fd = open(config_path, 'r')
+	data = fd.read()
+	fd.close()
```
Let us move these lines down into the try statement. 
Because, we dont need to open the config file at this point. If any of the following conditions return false, then it would be waste to read the config here.

kishore-sn commented on tools/dump_tool/ramdumpParser.py:
```diff
@@ -618,6 +622,12 @@ def main():
 
 
 	try:
+		# Get arch family
```
Read config file here. And remove the same logic from line 710/720.

tizen-build commented:
Target : [fa50db5d221c365076f1d734076c35da1b0ade47] - Code Rule Check OK.

deepaksrma commented on tools/debug/debugsymbolviewer.py:
```diff
@@ -0,0 +1,176 @@
+import re
+
+systemmap_path = '../../build/output/bin/System.map'	# System map file
+log_file = '../dump_tool/log_file'				# Crash log file
```
can you keep comment aligned?

sunghan-chang commented on build/configs/artik053/scripts/flash.ld:
```diff
@@ -72,6 +72,7 @@ SECTIONS
 
 	.text : {
 		_stext = ABSOLUTE(.);
+		_ram_kernel_text_start = ABSOLUTE(.);
```
1. I don't think this is good way. Because it requires unnecessary and wrong code. (This is not ram kernel text region)
How about adding new config?
When we enable the new config, then we check _ram_kernel_text_start.

2. How about renaming this like `_stext`? `_stext_ram`? I want to use the same policy for them.

sunghan-chang commented on os/arch/arm/src/common/up_checkspace.c:
```diff
@@ -32,17 +32,32 @@
+bool is_kernel_text_space(void *addr)
+{
+	/* Check if the address lies in the kernel text section range */
+	if ((addr >= (void *)&_stext && addr <= (void *)&_etext)
```
When new config is disabled, `)` is missing.

sunghan-chang commented on os/arch/arm/src/common/up_internal.h:
```diff
@@ -255,8 +255,10 @@ EXTERN uint32_t g_dabtstackbase;	/* Initial top of data abort stack */
-EXTERN uint32_t _etext;			/* End_1 of .text + .rodata */
+EXTERN uint32_t _stext_flash;			/* Start of .text in Flash region */
+EXTERN uint32_t _etext_flash;			/* End_1 of .text + .rodata in Flash region */
+EXTERN uint32_t _stext_ram;			/* Start of .text in RAM region */
```
Don't you need to cover ram variables with new config?

sunghan-chang commented on os/board/Kconfig:
```diff
@@ -191,6 +192,12 @@ config ARCH_IRQBUTTONS
 		"Support interrupts on button presses and releases."
 
+config ARCH_HAVE_RAM_KERNEL_TEXT
+	bool "Kernel text in RAM region"
```
As I recommend, this config should not be controlled by user. And naming `ARCH_HAVE_XXX` means architecture controls the config. Please remove `"Kernel text in RAM region"`.

sunghan-chang commented on os/drivers/kernel_test/kernel_test_drv.c:
```diff
@@ -142,7 +142,7 @@ static int kernel_test_drv_ioctl(FAR struct file *filep, int cmd, unsigned long
 		switch (obj->type) {
 		case MPUTEST_KERNEL_CODE:
-			obj->addr = &_stext;
+			obj->addr = &_stextflash;
```
typo, `_stext_flash`

sunghan-chang commented on os/arch/arm/src/armv7-m/up_hardfault.c:
```diff
@@ -134,17 +135,16 @@ int up_hardfault(int irq, FAR void *context, FAR void *arg)
 
-	if (((uintptr_t)pc >= (uintptr_t)&_stext && (uintptr_t)pc < (uintptr_t)&_etext) ||
-			(sched_self()->uspace &&
+	if ((is_kernel_text_space((void *)pc)) || (sched_self()->uspace &&
```
Is it wrong if we call `is_kernel_space` instead of `is_kernel_text_space`?

thapav replied:
I think these are a little different. 
*is_kernel_space* > Checks for text section, data & bss of kernel, whereas
*is_kernel_text_space* > Checks for only text section of kernel

sunghan-chang replied:
Yeah, the functions are different. But what I want to know is "do we need to use them separately in our usages?".

sunghan-chang replied:
If we call is_kernel_space instead of is_kernel_text_space, is there any problem?

thapav replied:
That is not an issue. Just to limit no. of checks that are being used in hard fault codes & for better code organization this is a better choice.

sunghan-chang commented:
1. Could you check conflict?
2. Please resolve build break due to miss `up_checkspace.c` in tiva/Make.defs.
```
LD: tinyara
/root/tizenrt/os/../build/output/libraries/libarch.a(up_hardfault.o): In function `up_hardfault':
/root/tizenrt/os/arch/arm/src/armv7-m/up_hardfault.c:147: undefined reference to `is_kernel_text_space'
Makefile:199: recipe for target '../build/output/bin/tinyara' failed
make[1]: *** [../build/output/bin/tinyara] Error 1
make[1]: Leaving directory '/root/tizenrt/os/arch/arm/src'
Makefile.unix:476: recipe for target 'pass2' failed
make: *** [pass2] Error 2
"qemu/build_test" ERROR!!!
```

tizen-build commented:
Target : [86b711118fdf0b420852ab72762467c29ca4a09a] - Code Rule Check OK.

sunghan-chang commented:
@thapav There is a conflict yet. Could you resolve it?

tizen-build commented:
Target : [6170a854a1fb61d179eb5729a88a73d573a6d29d] - Code Rule Check OK.

thapav commented:
> 1. Could you check conflict?
> 2. Please resolve build break due to miss `up_checkspace.c` in tiva/Make.defs.
> 
> ```
> LD: tinyara
> /root/tizenrt/os/../build/output/libraries/libarch.a(up_hardfault.o): In function `up_hardfault':
> /root/tizenrt/os/arch/arm/src/armv7-m/up_hardfault.c:147: undefined reference to `is_kernel_text_space'
> Makefile:199: recipe for target '../build/output/bin/tinyara' failed
> make[1]: *** [../build/output/bin/tinyara] Error 1
> make[1]: Leaving directory '/root/tizenrt/os/arch/arm/src'
> Makefile.unix:476: recipe for target 'pass2' failed
> make: *** [pass2] Error 2
> "qemu/build_test" ERROR!!!
> ```

Addressed the comments. Kindly check

jeongchanKim reviewed (approved):
LGTM

deepaksrma reviewed (commented):
LGTM

sunghan-chang commented on os/board/Kconfig:
```diff
@@ -191,6 +192,11 @@ config ARCH_IRQBUTTONS
 		"Support interrupts on button presses and releases."
 
+config ARCH_HAVE_RAM_KERNEL_TEXT
+	default n
```
Maybe this causes menuconfig error or warning.
```
config ARCH_HAVE_..
     bool
     default n
     ---help---
           To support ...
````
You should do like above.

thapav replied:
done

tizen-build commented:
Target : [6515ac4f0cf841552102a98d3d03af2cebda348e] - Code Rule Check OK.

tizen-build commented:
Target : [dd4dc5561d7dbb9995f78c286def2dd337b0c882] - Code Rule Check OK.

sunghan-chang reviewed (commented):
Could you modify nettest's defconfig as well?

sunghan-chang commented on os/include/tinyara/arch.h:
```diff
@@ -2179,6 +2179,24 @@ void up_wdog_keepalive(void);
+ *   Check the address is in kernel data space or not
+ *
+ ****************************************************************************/
+bool is_kernel_data_space(void *addr)
```
missing `;`

sunghan-chang commented on os/arch/arm/src/common/up_checkspace.c:
```diff
@@ -32,17 +32,43 @@
+bool is_kernel_data_space(void *addr)
+{
+	/* Check if the address lies in the kernel data section range */
+	if (addr >= (void *)&_sdata && addr <= (void *)&_ebss) {
```
This is valid when bss is located at next of data section. But can we guarantee that or should we layout like that always? I don't think so. So, how about separating bss and data?
```c
if ((addr >= (void *)&_sdata && addr <= (void *)&_edata) || (addr >= (void *)&_sbss && addr <= (void *)&_ebss)) {
```

tizen-build commented:
Target : [229acf166487cf1450b5f03f9bedd3ea4359e323] - Code Rule Check OK.

tizen-build commented:
Target : [a3c5b1707aad773bfb52f9bb501c26cfae0a1165] - Code Rule Check OK.

sunghan-chang reviewed (approved):
Once you confirm operation with latest changes, I will merge this PR. Please let me know test result.

thapav commented:
A sample tool run >>
`
*************************************************************
dump_file         : None
log_file          : ./log_file
Tinyara_elf_file  : ../../build/output/bin/tinyara.axf
*************************************************************

----------------------------------  DEBUG DISPLAY SYMBOLS -------------------------------
Dump_address     Symbol_address   Symbol_name   File_name
0xe0054f1        0xe0054a4        up_assert     /root/tizenrt/os/include/assert.h:211
0xe0532d5        0xe05326c        __FUNCTION__.5946
0xe05325f        0xe05324f        __FUNCTION__.5923
0xe053d32        0xe053ca8        __FUNCTION__.5294
0xe006049        0xe00603c        hello_main    /root/tizenrt/apps/examples/hello/hello_main.c:71
0xe00603d        0xe00603c        hello_main    /root/tizenrt/apps/examples/hello/hello_main.c:71
0xe001b2d        0xe001b00        task_start    /root/tizenrt/os/kernel/task/task_start.c:133
0xe006049        0xe00603c        hello_main    /root/tizenrt/apps/examples/hello/hello_main.c:71
0xe00603d        0xe00603c        hello_main    /root/tizenrt/apps/examples/hello/hello_main.c:71
0xe001b2d        0xe001b00        task_start    /root/tizenrt/os/kernel/task/task_start.c:133

PC_value         Symbol_address   Symbol_name   File_name
0xe0054a4        0xe0054a4        up_assert     /root/tizenrt/os/include/assert.h:211
-----------------------------------------------------------------------------------------
`

Added registerdump value manually in logs for verification.
